source/validate_decorations.cpp

```diff
@@ -59,6 +60,298 @@ bool hasImportLinkageAttribute(uint32_t id, ValidationState_t& vstate) {
+// Offset decorations placing its first byte at F and its last byte at L, where
+// floor(F / 16) != floor(L / 16).
+// - It is a vector with total size greater than 16 bytes and has its Offset
+// decorations placing its first byte at a non-integer multiple of 16.
+bool hasImproperStraddle(uint32_t id, uint32_t offset,
+                         ValidationState_t& vstate) {
+  const auto inst = vstate.FindDef(id);
+  const auto words = inst->words();
+  const auto size = getSize(id, false, vstate);
+  const auto F = offset;
+  const auto L = offset + size - 1;
+  const auto floorF = floor((float)F / 16.0f);
+  const auto floorL = floor((float)L / 16.0f);
+  if (size <= 16) {
+    if (floorF != floorL) return true;
+  } else {
+    if (F % 16 != 0) return true;
+  }
+  return false;
+}
+
+// Check alignment of x. For storage buffers the alignment can be either rounded
+// up or not.
+bool checkAlignment(uint32_t x, uint32_t alignment, uint32_t alignmentRoundedUp,
+                    bool isBlock) {
+  if (isBlock) {
+    if (alignmentRoundedUp == 0) {
+      if (x != 0) return false;
+    } else if (x % alignmentRoundedUp != 0)
+      return false;
+  } else {
+    if (alignment == 0 || alignmentRoundedUp == 0) {
+      if (x != 0) return false;
+    } else if (x % alignment != 0 && x % alignmentRoundedUp != 0)
+      return false;
+  }
+  return true;
+}
+
+// Checks for standard layout rules.
+bool checkLayout(uint32_t struct_id, bool isBlock, ValidationState_t& vstate) {
+  if (vstate.options()->relax_block_layout) return true;
+  const auto members = getStructMembers(struct_id, vstate);
+  uint32_t padStart = 0, padEnd = 0;
+  for (size_t memberIdx = 0; memberIdx < members.size(); memberIdx++) {
+    auto id = members[memberIdx];
+    const auto baseAlignment = getBaseAlignment(id, false, vstate);
+    const auto baseAlignmentRoundedUp = getBaseAlignment(id, true, vstate);
+    const auto inst = vstate.FindDef(id);
+    const auto opcode = inst->opcode();
+    uint32_t offset = 0xffffffff;
+    for (auto& decoration : vstate.id_decorations(struct_id)) {
+      if (SpvDecorationOffset == decoration.dec_type() &&
+          decoration.struct_member_index() == (int)memberIdx) {
+        offset = decoration.params()[0];
+      }
+    }
+    const auto size = getSize(id, isBlock, vstate);
+    const auto lastByte = size - 1;
+    // Check offset.
+    if (offset == 0xffffffff) return false;
+    if (!checkAlignment(offset, baseAlignment, baseAlignmentRoundedUp, isBlock))
+      return false;
+    if (offset >= padStart && offset + lastByte <= padEnd) return false;
+    // Check improper straddle of vectors.
+    if (SpvOpTypeVector == opcode && hasImproperStraddle(id, offset, vstate))
+      return false;
+    // Check struct members recursively.
+    if (SpvOpTypeStruct == opcode && !checkLayout(id, isBlock, vstate))
+      return false;
+    // Check matrix stride.
+    if (SpvOpTypeMatrix == opcode) {
```

Comment:
This is looking for a MatrixStride decoration on the *matrix type*. 
But the MatrixStride decoration decoration is on the struct member.  So this will not trigger.

I think this is the root cause of something I'm seeing now, i.e. that MatrixStride itself is not being validated.